rp2040-hal/src/spi.rs

```diff
@@ -78,53 +78,131 @@ impl<S: State, D: SpiDevice, const DS: u8> Spi<S, D, DS> {
-                postdiv = postdiv_option;
-                break;
-            }
-        }
+        debug_assert_ne!(settings.prescale, u8::MAX);
+
+        self.set_baudrate_from_settings(&settings);
+
+        // Return the frequency we were able to achieve
+        use fugit::RateExtU32;
+        (freq_in / (settings.prescale as u32 * (1 + settings.postdiv as u32))).Hz()
+    }
 
+    /// Set the baudrate using a previously calculated [SpiClockDividerSettings]
+    pub fn set_baudrate_from_settings(&mut self, settings: &SpiClockDividerSettings) {
         self.device
             .sspcpsr
-            .write(|w| unsafe { w.cpsdvsr().bits(prescale) });
+            .write(|w| unsafe { w.cpsdvsr().bits(settings.prescale) });
         self.device
             .sspcr0
-            .modify(|_, w| unsafe { w.scr().bits(postdiv) });
+            .modify(|_, w| unsafe { w.scr().bits(settings.postdiv) });
+    }
 
-        // Return the frequency we were able to achieve
-        use fugit::RateExtU32;
-        (freq_in / (prescale as u32 * (1 + postdiv as u32))).Hz()
+    /// Set the mode
+    ///
+    /// Momentarily disables / enables the device so be careful of truncating ongoing transfers.
+    pub fn set_mode(&mut self, mode: Mode) {
+        // disable the device
+        self.device.sspcr1.modify(|_, w| w.sse().clear_bit());
+
+        // Set the polarity and phase
+        self.device.sspcr0.modify(|_, w| {
+            w.spo()
+                .bit(mode.polarity == Polarity::IdleHigh)
+                .sph()
+                .bit(mode.phase == Phase::CaptureOnSecondTransition)
+        });
+
+        // enable the device
+        self.device.sspcr1.modify(|_, w| w.sse().set_bit());
+    }
+
+    /// Wait until all operations have completed and the bus is idle.
+    pub fn flush(&self) -> Result<(), nb::Error<Infallible>> {
+        if self.device.sspsr.read().bsy().bit() {
+            Err(nb::Error::WouldBlock)
+        } else {
+            Ok(())
+        }
     }
 }
 
+/// Clock divider settings
+pub struct SpiClockDividerSettings {
+    /// The prescaler for writing to sspcpsr
+    pub prescale: u8,
+    /// The postdiv for writing to sspcr0
+    pub postdiv: u8,
+}
+
+/// Calculate the prescale and post divider settings for a required baudrate that can then be
+/// passed to [Spi::set_baudrate_from_settings]
+///
+/// This calculation takes ~100us on rp2040 at runtime and is used by [set_baudrate] every time
+/// it's called. That might not be acceptable in
+/// situations where you need to change the baudrate often.
+///
+/// Note that this is a const function so you can use it in a static context if you don't change your
+/// peripheral clock frequency at runtime.
+///
+/// If you do change your peripheral clock at runtime you can store the [SpiClockDividerSettings] and only re-calculate it
+/// when the peripheral clock frequency changes.
+pub const fn calc_spi_clock_divider_settings_for_baudrate(
```

Comment:
I think this would make a nice inherent method to `SpiClockDividerSettings` like

```rust
impl SpiClockDividerSettings {
    pub const fn new(peri_frequency: HertzU32, baudrate: HertzU32) -> Self {
        todo!()
    }
}
```